tests/utilities/nidigital_measurement/__init__.py

```diff
@@ -38,12 +39,20 @@ def measure(
                 connections = reservation.get_nidigital_connections(pin_names)
                 assert all([session_info is not None for session_info in session_infos])
                 passing_sites, failing_sites = _burst_spi_pattern(session_infos)
+                site0_connections, site1_connections = [], []
+                # Assumption: All pins connected to site0 belongs to one session and
```

Comment:
If you are trying to group the channels by session, then why not do exactly that?

Grouping the channels by site would be almost the same, but the key function would use `conn.site` instead of `conn.session_info.session_name`. However, the comment suggests you aren't really trying to group by site.

Yes, `sorted` is stable.

```python
from itertools import groupby

                key_func = lambda conn: conn.session_info.session_name
                connections_by_session = [
                    list(g) for _, g in groupby(sorted(connections, key=key_func), key=key_func)
                ]

                return (
                    ...
                    [
                        ", ".join(conn.channel_name for conn in conns)
                        for conns in connections_by_session
                    ],
                    ...
                )

Reply:
Yeah, Thank you. I've updated the construction logic as per your suggestion.

Reply:
Ok, this is fine.

> maybe they missed handling this special case.

They explicitly chose not to handle this case: https://github.com/python/mypy/blob/bc591c756a453bb6a78a31e734b1f0aa475e90e0/mypy/renaming.py#L431

```python
    If there's a reference to a variable defined in 'with' outside the
    statement, or if there's any trickiness around variable visibility
    (e.g. function definitions), we give up and won't perform renaming.
```